tso-server/server/server.go

```diff
@@ -41,46 +49,110 @@ type atomicObject struct {
 
 // TimestampOracle generates a unique timestamp.
 type TimestampOracle struct {
-	ts     atomic.Value
-	ticker *time.Ticker
+	cfg *Config
+
+	ts atomic.Value
 
 	listener net.Listener
 
 	wg       sync.WaitGroup
 	connLock sync.Mutex
 	conns    map[net.Conn]struct{}
+
+	zkElector *zkhelper.ZElector
+	zkConn    zkhelper.Conn
+
+	isLeader int64
+
+	lastSavedTime time.Time
 }
 
-func (tso *TimestampOracle) updateTicker() {
-	// TODO: save latest TS to persistent storage(zookeeper/etcd...)
+func (tso *TimestampOracle) loadTimestamp() error {
+	last, err := loadTimestamp(tso.zkConn, tso.cfg.RootPath)
+	if zkhelper.ZkErrorEqual(err, zk.ErrNoNode) {
+		// no timestamp node, create later
+		err = nil
+	}
+
+	if err != nil {
+		return errors.Trace(err)
+	}
+
+	var now time.Time
+
 	for {
-		select {
-		case <-tso.ticker.C:
-			prev := tso.ts.Load().(*atomicObject)
-			now := time.Now()
-
-			// ms
-			since := now.Sub(prev.physical).Nanoseconds() / 1e6
-			if since > 2*updateTimtestampStep {
-				log.Warnf("clock offset: %v, prev:%v, now %v", since, prev.physical, now)
-			}
-			// Avoid the same physical time stamp
-			if since <= 0 {
-				log.Warn("invalid physical time stamp")
-				continue
-			}
+		now = time.Now()
 
-			current := &atomicObject{
-				physical: now,
-			}
-			tso.ts.Store(current)
+		since := (now.UnixNano() - last) / 1e6
+		if since <= 0 {
+			return errors.Errorf("%s <= last saved time %s", now, time.Unix(0, last))
```

Comment:
Maybe we should interrupt the task if server's time is delayed too much. Because in that case the server would never become the leader successfully, and it's hard for user to notice the situation.